processing/src/main/java/org/apache/druid/math/expr/ConstantExpr.java

```diff
@@ -464,15 +473,18 @@ public String toString()
 
 class ComplexExpr extends ConstantExpr<Object>
 {
+  final ExprEval expr;
```

Comment:
Curious since we are evaluating this early, does the equals and hash method need to honor this ?

Reply:
its not included in equals / hashCode as its a computed value from the `value` constructor argument;

as a matter of fact... `ConstantExpr` also has an `outputType` field which is also ignored in the `equals` / `hashCode` stuff - as that is passed in as a constructor parameter - actually this could probably be refactored to remove that field; however I'm not sure what benefits that would give...

the other way around could be to build the `ContantExpr` on top of an `ExprEval` - and use `ExprEval.value()` to get the inner value - but that approach would not like how the current `BigIntegerExpr` [works](https://github.com/apache/druid/blob/d5b1b5bc8e751b8a45b2a2bf5a02ee2018968a9f/processing/src/main/java/org/apache/druid/math/expr/ConstantExpr.java#L126)...

Reply:
I'm not sure how often `Expr` are used in equals/hashcode, my gut says not often since the thing that makes them also usually has the original string which was parsed into `Expr` and is likely cheaper to just use the string